src/keyszer/config_api.py

```diff
@@ -213,17 +218,21 @@ def unicode_keystrokes(n):
     """Turn Unicode number into keystroke commands"""
     if (n > 0x10ffff):
         raise UnicodeNumberToolarge(f"{hex(n)} too large for Unicode keyboard entry.")
-
-    combo_list = [
-        combo("Shift-Ctrl-u"),
-        *[Key[hexdigit]
-            for digit in _digits(n, 16)
-            for hexdigit in hex(digit)[2:].upper()
-            ],
-        Key.ENTER,
-    ]
-
-    return combo_list
+    def _unicode_keystrokes(ctx):
+        combo_list = [
+            combo("Shift-Ctrl-u"),  # requires "ibus" or "fctix" as input manager?
+            *[Key[hexdigit]
+                for digit in _digits(n, 16)
+                for hexdigit in hex(digit)[2:].upper()
+                ],
+            Key.ENTER,
+        ]
+        if ctx.capslock_on:
+            combo_list.insert(0, Key.CAPSLOCK)
+            combo_list.append(Key.CAPSLOCK)
```

Comment:
Would this approach also work for us_keystrokes?  Would it be better/worse?

Reply:
I don't think so. I kind of tried that first. The situation is different between the two functions. The Unicode function is used by the string processor when it needs to (when a character is outside the ASCII range) but it can also be used independently. For the Unicode sequence, it only matters that you disable the CapsLock LED before attempting to perform the Unicode entry shortcut, so that it doesn't interfere with the shortcut, and then you need to re-enable it after the sequence is over. Obviously, so nothing changes for the user as they type. 

For the string processor, the CapsLock LED state affects only the letter keys, not punctuation or numbers or anything else. So you really need to get in there and just invert the state of "Shifting" only the keys that are affected by the CapsLock state. 

I believe the way it is set up is just about the best possible way to solve the issue. It seems to work very well, regardless of whether you are just processing a string with or without any Unicode/extended characters, or directly using the Unicode processor function. 

I think one of the benefits here is that we don't actually ever toggle the CapsLock LED if there are no non-ASCII characters in the string. We just "adapt" to its state when that is feasible. That was not an option in the Unicode function. Trying to invert the "Shifting" would not fix the shortcut.